src/lib/Imp.hs

```diff
@@ -1396,7 +1398,20 @@ computeOffset idxNest' idxs = do
 sumUsingPolysImp :: Emits n => SIAtom n -> Abs (Binder SimpToImpIR) SIBlock n -> SBuilderM n (SIAtom n)
 sumUsingPolysImp lim (Abs i body) = do
   ab <- hoistDecls i body
-  sumUsingPolys lim ab
+  PairE lim' ab' <- emitImpVars (PairE lim ab) -- Algebra only works with Atom vars
+  sumUsingPolys lim' ab'
+
+emitImpVars ::
+  (Emits n, HoistableE e, SubstE (SubstVal ImpNameC SIAtom) e, SinkableE e, SubstE Name e)
+  => e n -> SBuilderM n (e n)
+emitImpVars e = do
+  let impVars = nameSetToList (freeVarsE e) :: [ImpName _]
+  Abs impBs e' <- return $ abstractFreeVarsNoAnn impVars e
+  substVals <- forM impVars \v -> do
+    ty <- impVarType v
+    atomVar <- emitAtomToName (getNameHint v) (AtomicIVar (LeftE v) ty)
+    return $ (SubstVal (Var atomVar) :: SubstVal ImpNameC SIAtom ImpNameC _)
+  applySubst (impBs @@> substVals) e'
```

Comment:
Why do we never have to undo this imp->core variable change?

Reply:
This is in the (Simp) builder monad. We bring things back to Imp by translating the resulting Simp block to an Imp block.